File: C++/Array/Best Time to Buy and Sell Stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        
        int n = prices.size();
        int maxRight = prices[n-1];
        int maxProfit = 0;
        for(int i = n - 2; i >= 0; i--){
            maxRight = max(prices[i], maxRight);
            maxProfit = max(maxProfit, maxRight - prices[i]);
        }
        
        return maxProfit;
        
    }
};

File: Python/Array/Best Time to Buy and Sell Stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # stack approach - Next greater element to the right
        # O(n)
        if len(prices)==0:
            return 0
            
        l=[-1]*len(prices)
        max_val=prices[len(prices)-1]
        for i in range(len(prices)-2, -1,-1):
            if max_val>prices[i]:
                l[i]=max_val
            else:
                max_val=max(max_val,prices[i])
        max_profit=float("-inf")        
        for i in range(len(prices)):
            if l[i]!=-1:
                max_profit=max(max_profit,l[i]-prices[i])
        if max_profit==float("-inf"):
            return 0
        return max_profit
